Simplify function enter macros for performance benefits

gnuoyd opened the pull request:
Perform library/package initialization explicitly instead of testing at the top of every function whether the library/package is initialized and performing initialization there. This knocks 11% off of the user time spent by the benchmark used in HDFFV-9671.

This isn't quite finished. Some new warnings have cropped up with the macro changes. I'll update the PR as I quiet the warnings.

gnuoyd committed:
Take a stab at using constructors to initialize instead of
function-entry macros.  This is a work in progress.  It's good enough to
run `many_dsets`.

github-actions[bot] committed:
Committing clang-format changes

gnuoyd committed:
Add the `many_dsets` benchmark and some scripts I used on jelly for
setting up the build/test environment and for recording/flame-graphing
profiles.

github-actions[bot] committed:
Committing clang-format changes

gnuoyd committed:
Change my Makefile and environment script to work both on jelly
and on mayll (and probably on Summit).

gnuoyd committed:
Merge branch 'manydsets-start' into manydsets-constructors

gnuoyd committed:
Disable clang-format "fix."

gnuoyd committed:
Merge branch 'manydsets-start' into manydsets-constructors

gnuoyd committed:
Replace the `if (!H5_TERM_GLOBAL)` test in each FUNC_ENTER_ macro with
`if (true)`.

gnuoyd committed:
Fix bad grammar in a comment.

gnuoyd committed:
Instead of labeling the H5*__init_package routines constructors, fold
each into an initialization routine, H5*_init(), and call each of
the H5*_init() routines.  Call most of the H5*_init() routines from
H5_init_library() in an explicit order that I found out earlier by
instrumenting each __init_package routine and running the library
tests.  Roll H5FD*__init_package routines into H5FD*_init() routines.
This change ends just-in-time initialization of package dependencies by
package initializers.

Don't track in per-package variables (H5_PKG_INIT_VAR) whether each
package has been initialized.  Instead, track in a single library
variable whether the whole library is initialized or not.

Drive the initialization of packages by H5_init_library() with a table
of initializer routines.  Also drive the termination of packages by
H5_term_library() with a table.

Perform initialization as needed from FUNC_ENTER_API_INIT(err).  This
basically restores the old behavior of that macro.

Delete a bunch of #definitions in H5private.h that have fallen out of
use with these changes.

github-actions[bot] committed:
Committing clang-format changes

gnuoyd committed:
Undo the bad auto-formatting that appears to have occurred in spite
of my disabling it.  Bracket some code in /* clang-format off */ /*
clang-format on */ to prevent a recurrence.

gnuoyd committed:
Remove a diagnostic abort().

gnuoyd committed:
Fix a logic error: print a comma between every package terminator run,
and don't print an initial comma.

gnuoyd committed:
Complete the changes I started in H5_term_library() that undo the bad
auto-formatting.

Stop tracking whether package "tops" were initialized in per-package
variables H5*_top_package_initialize_s.  H5_term_library() takes care of
that for them.

Remove H5R_top_term_package() and H5R_term_package(), they don't do
anything.

github-actions[bot] committed:
Committing clang-format changes

gnuoyd committed:
NFCI.  Simplify macro text: replace `if (true) {` with `{`.

gnuoyd committed:
Fix formatting and suppress clang-format on a longer range.

gnuoyd committed:
Quiet some unused label, unused variable complaints that cropped up
after I simplified the FUNC_ENTER_ macros for the sake of performance.

github-actions[bot] committed:
Committing clang-format changes

gnuoyd committed:
Merge remote-tracking branch 'thg/develop' into manydsets-start

gnuoyd committed:
Merge branch 'manydsets-start' into manydsets-total-quick-enter

gnuoyd committed:
Merge branch 'manydsets-total-quick-enter' of github.com:gnuoyd/hdf5 into manydsets-total-quick-enter

gnuoyd committed:
Delete some programs and scripts that don't belong in the pull request.

gnuoyd committed:
Use the right function-entry macro.

gnuoyd committed:
Use a sensible format and disable auto-formatting.

gnuoyd committed:
Stop calling do-nothing initializer H5FS_init().  Delete it.

gnuoyd committed:
Document what changes to make if the default VFD changes.

gnuoyd committed:
While I am here, change an `await_prior` flag on the terminator table
to `true` to match the previous, non-table-driven code that was here.
Found the oversight making the following changes:

NFCI: insert an empty line and copy over slightly-edited comments from
the previous version, where those comments still correctly explained how
library termination operated.

gnuoyd committed:
NFCI: lower a staircase.

gnuoyd committed:
Replace every occurrence of FUNC_ENTER_NOAPI_INIT(...) with H5_PUSH_FUNC
since that is all that that macro does any more.

Quiet a bunch of new warnings by changing FUNC_ENTER_NOAPI(...) to
FUNC_ENTER_NOAPI_NOERR and removing disused `done:` labels.

gnuoyd committed:
NFCI: add curly braces around a multiline statement.

gnuoyd committed:
Quiet a signed/unsigned comparison warning.

gnuoyd committed:
Add some documentation about library initialization and shutdown.

gnuoyd committed:
Make sure that the library is initialized, or else that initialization
is already underway, before performing any VFD's initialization.

github-actions[bot] committed:
Committing clang-format changes

gnuoyd committed:
Merge remote-tracking branch 'thg/develop' into manydsets-total-quick-enter

github-actions[bot] committed:
Committing clang-format changes

gnuoyd committed:
Reduce differences from `develop` branch.

gnuoyd committed:
Always initialize `tot_init`.

github-actions[bot] committed:
Committing clang-format changes

gnuoyd committed:
Merge remote-tracking branch 'thg/develop' into manydsets-total-quick-enter

gnuoyd commented:
This is ready for review. @derobins, please have a look.

soumagne commented on src/H5.c:
```diff
@@ -145,8 +136,13 @@ H5__init_package(void)
 
+    /* Run the library initialization routine, if it hasn't already run */
+    if (H5_INIT_GLOBAL || H5_TERM_GLOBAL)
+        HGOTO_DONE(SUCCEED)
```
does that mean that if H5_term_global is set there's no need to initialize anymore ? it looks like once you finalized you can't initialize again? or it's no longer needed ?

gnuoyd replied:
H5_TERM_GLOBAL is set before library termination and reset after, so the library can reinitialize.

The check in H5_init_library ensures that the library cannot start to reinitialize while it is terminating. In the New Initialization Order, there are fewer ways to enter initialization, so we may be able to eliminate the check sometime in the future.

soumagne commented on src/H5.c:
```diff
@@ -257,26 +253,32 @@ H5_init_library(void)
+    /* clang-format off */
+    struct {
+        herr_t (*func)(void);
+        const char *descr;
```
It would be nice I think to have a macro like you did for TERMINATOR that combines both phases of the initialization, e.g., H5VL_init_phase1 and H5VL_init_phase2 for "VOL"

gnuoyd replied:
I don't understand what you mean.

soumagne replied:
Just meant to have 2 funcs instead of one and a macro that sets the phase1/phase2 but not a big deal, just a suggestion. Just so that you have only one line for each module instead of having it decomposed.

gnuoyd replied:
Let's revisit this later if more phased initializations crop up.

soumagne commented on src/H5.c:
```diff
@@ -257,26 +253,32 @@ H5_init_library(void)
+    } initializer[] = {
+        {H5E_init, "error"}
+    ,   {H5VL_init_phase1, "VOL"}
+    ,   {H5SL_init, "VOL"}
```
desc for H5SL_init is VOL ?

gnuoyd replied:
Typo. Will fix.

soumagne commented on src/H5FDcore.c:
```diff
@@ -491,10 +458,30 @@ H5FD__init_package(void)
+     */
+    if (H5_init_library() < 0) {
+        HGOTO_ERROR(H5E_FUNC, H5E_CANTINIT, H5I_INVALID_HID, "library initialization failed")
+    }
```
It's relatively annoying I think that we have to add that `H5_init_library()` here and there in the library... it would be good to find a way to get rid of that to make sure nobody eventually forgets one... it's not that clear to me why that's actually needed here

gnuoyd replied:
An application can enter the library for the first time by evaluating one of the H5FD_<vfd> symbols such as H5FD_CORE. These H5_init_library() calls are there to make sure that the library is initialized before the VFD initialization proceeds.

gnuoyd replied:
One thing that I could do is create a wrapper function, `hid_t H5FD_perform_init(hid_t (*init)(void))`, that does `H5_init_library`, checks its status, and on success calls `init`. Then each of the `H5FD_<vfd>` symbols can be `#define H5FD_<vfd> H5FD_perform_init(H5FD_<vfd>_init)`.

Right now some VFDs are effectively "internal" to the library and others are "external". The external ones cannot "see" H5_init_library and other library-internal functions, so they make sure the library is initialized with a call to H5open. H5FD_perform_init() can smooth over that difference.

soumagne replied:
Yes anything that would make things easier would be good. One concern that I have for now is for external VFDs and VOLs maintainers, that init mechanism should be as transparent as possible for them.

derobins replied:
We could reduce a lot of complexity if we simply required people to explicitly initialize the library, as in MPI with MPI_Init().

soumagne replied:
right or we could also just simply use constructor attributes since we don't use any args for init. Would also be cleaner than the atexit which would become a destructor.

gnuoyd replied:
My most recent change introduces `hid_t H5FDperform_init(hid_t (*init)(void))` and uses it for the H5FD_<vfd> symbols. Reduces duplication nicely.

soumagne replied:
definitely looks nicer I think.

soumagne commented on src/H5T.c:
```diff
@@ -801,7 +759,10 @@ H5T__init_package(void)
+    FUNC_ENTER_NOAPI(FAIL)
+
+    if (H5_TERM_GLOBAL)
+        HGOTO_DONE(SUCCEED)
```
Why does this one has the check for H5_TERM_GLOBAL, it didn't seem like the other packages needed it

gnuoyd replied:
I will have another look. It may be some oddity that I copied from the previous code without dwelling on it, or it may be something that slipped in during a previous iteration of the patch.

gnuoyd replied:
Removing it didn't seem to hurt anything.

derobins commented on src/H5.c:
```diff
@@ -101,32 +100,24 @@ static H5_atclose_node_t *H5_atclose_head = NULL;
 
+    if (id == H5I_INVALID_HID) {
+        HGOTO_ERROR(H5E_FUNC, H5E_CANTINIT, FAIL, "unable to load default VFD ID")
+    }
```
The library idiom is to combine the assignment and the check on the same line. It's a longer code line, but it ensures that values don't slip by unchecked.

gnuoyd replied:
No intermediate variable is necessary here.

derobins commented on src/H5.c:
```diff
@@ -101,32 +100,24 @@ static H5_atclose_node_t *H5_atclose_head = NULL;
-    Non-negative on success/Negative on failure
-DESCRIPTION
-    Initializes any interface-specific data or routines.
---------------------------------------------------------------------------*/
```
There's a lot to dislike about our function comments, but please don't change the coding style without a plan to propagate the new paradigm throughout the library.

gnuoyd replied:
I have added a function comment I found elsewhere to _init routines that were missing them.

derobins commented on src/H5.c:
```diff
@@ -346,108 +350,137 @@ H5_term_library(void)
+            if (nprinted < 0)
+                continue;
+            if ((size_t)nprinted >= nleft)
+                nprinted = snprintf(next, nleft, "...");
```
HD prefixes on POSIX and std C calls. Please don't change the coding standard unilaterally.

gnuoyd committed:
Fix typo: H5SL_init initializes skip lists, not VOL.

gnuoyd committed:
Remove H5_TERM_GLOBAL test in H5T_init.  H5T_init was unusual in that
it tested H5_TERM_GLOBAL and exited early if it was set.  No other
module initializers did that, and I cannot find any reason that should
be necessary.  Tests still pass when I remove it, so away it goes.

gnuoyd committed:
Use HD prefix.

gnuoyd committed:
Add function header comments.

gnuoyd committed:
Drop the intermediate variable, it's only used once.

epourmal commented:
It would be good, but we will have problems with existing apps, right?

Sent from my iPhone

On Nov 3, 2021, at 6:05 PM, Dana Robinson ***@***.***> wrote:

﻿

@derobins commented on this pull request.

________________________________

In src/H5FDcore.c<https://github.com/HDFGroup/hdf5/pull/1024#discussion_r742411482>:

>
     FUNC_ENTER_NOAPI(H5I_INVALID_HID)

+    /* It is possible that an application will call this routine through
+     * a `H5FD_*` symbol (`H5FD_FAMILY`, `H5FD_MULTI`, `H5FD_SEC2`, et
+     * cetera) before the library has had an opportunity to initialize.
+     * Call H5_init_library() to make sure that the library has been
+     * initialized before this VFD tries to initialize.
+     */
+    if (H5_init_library() < 0) {
+        HGOTO_ERROR(H5E_FUNC, H5E_CANTINIT, H5I_INVALID_HID, "library initialization failed")
+    }


We could reduce a lot of complexity if we simply required people to explicitly initialize the library, as in MPI with MPI_Init().

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/HDFGroup/hdf5/pull/1024#discussion_r742411482>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADLFT3LMMAHTWS7ASXMBJBLUKHE3NANCNFSM5EIF7QPA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

gnuoyd committed:
Extract subroutine `H5FDperform_init(hid_t (*init)(void))` that
initializes the library, if necessary, before calling its VFD-initializer
argument.  Use H5FDperform_init in the definition of the symbols
H5FD_<vfd> (e.g., H5FD_SEC2), which may be evaluated before the library
is initialized, like so:

```
```

I implement H5FDperform_init in its own source file, H5FDperform.c,
and exclude that file from trace processing because the `bin/trace`
cannot deal with the function-pointer type.

gnuoyd committed:
Straggler from last: add new source file src/H5FDperform.c.

github-actions[bot] committed:
Committing clang-format changes

derobins commented on src/H5.c:
```diff
@@ -346,108 +356,137 @@ H5_term_library(void)
+        bool        completed;   /* true iff this terminator was already
+                                  * completed
+                                  */
+        const bool  await_prior; /* true iff all prior terminators in the list
```
Internally we should be using hbool_t for consistency. I'd like to strip hbool_t from the library, but until then we should continue to use it.

derobins commented on src/H5FDsec2.c:
```diff
@@ -233,10 +200,20 @@ H5FD__init_package(void)
+        ignore_disabled_file_locks_s = FALSE; /* Override: Don't ignore disabled locks */
+    else
+        ignore_disabled_file_locks_s = FAIL; /* Environment variable not set, or not set correctly */
+
```
I'm making a mental note to put this code block in a function...

derobins commented:
If you can, please switch that one bool to hbool_t. I've been trying to prioritize removing hbool_t since it's unnecessary w/ C99, but it's a big job and I haven't had the time.

derobins commented:
> It would be good, but we will have problems with existing apps, right?
> Sent from my iPhone On Nov 3, 2021, at 6:05 PM, Dana Robinson ***@***.***> wrote: ﻿ @derobins commented on this pull request.
> ________________________________ In src/H5FDcore.c<[#1024 (comment)](https://github.com/HDFGroup/hdf5/pull/1024#discussion_r742411482)>:
> FUNC_ENTER_NOAPI(H5I_INVALID_HID) + /* It is possible that an application will call this routine through + * a `H5FD_*` symbol (`H5FD_FAMILY`, `H5FD_MULTI`, `H5FD_SEC2`, et + * cetera) before the library has had an opportunity to initialize. + * Call H5_init_library() to make sure that the library has been + * initialized before this VFD tries to initialize. + */ + if (H5_init_library() < 0) { + HGOTO_ERROR(H5E_FUNC, H5E_CANTINIT, H5I_INVALID_HID, "library initialization failed") + } We could reduce a lot of complexity if we simply required people to explicitly initialize the library, as in MPI with MPI_Init(). — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub<[#1024 (comment)](https://github.com/HDFGroup/hdf5/pull/1024#discussion_r742411482)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADLFT3LMMAHTWS7ASXMBJBLUKHE3NANCNFSM5EIF7QPA>. Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

Yes. All applications would require changes.

gnuoyd committed:
Merge remote-tracking branch 'thg/develop' into manydsets-total-quick-enter

gnuoyd committed:
Add a missing file to the MANIFEST.

gnuoyd committed:
Switch to FUNC_ENTER_API_NOINIT in H5FDperform_init() and hbool_t in
H5_term_library().
